File: ruby/twelve-days/twelve_days.rb
#  frozen_string_literal: true

module TwelveDays
  CARDINALS =
    Hash.new { |hash, key| hash[key] = key.gsub(/th\z/, "") }.tap do |hash|
      hash["first"] = "a"
      hash["second"] = "two"
      hash["third"] = "three"
      hash["fifth"] = "five"
      hash["eighth"] = "eight"
      hash["ninth"] = "nine"
      hash["twelfth"] = "twelve"
    end
  private_constant :CARDINALS
  DECLARATION_OF_RECEIPT =
    "On the %<day>s day of Christmas my true love gave to me: "
  private_constant :DECLARATION_OF_RECEIPT
  # Empty song holds an array for the generated song verses, and an array for
  # the "extra gifts" that need to be added to each consecutive day.
  EMPTY_SONG = [[], []].freeze
  private_constant :EMPTY_SONG
  GIFTS = [
    " Partridge in a Pear Tree",
    " Turtle Doves",
    " French Hens",
    " Calling Birds",
    " Gold Rings",
    " Geese-a-Laying",
    " Swans-a-Swimming",
    " Maids-a-Milking",
    " Ladies Dancing",
    " Lords-a-Leaping",
    " Pipers Piping",
    " Drummers Drumming"
  ].freeze
  private_constant :GIFTS
  NEWLINE = "\n"
  private_constant :NEWLINE
  ORDINALS = %w[
    first
    second
    third
    fourth
    fifth
    sixth
    seventh
    eighth
    ninth
    tenth
    eleventh
    twelfth
  ].freeze
  private_constant :ORDINALS
  PERIOD = "."
  private_constant :PERIOD

  module_function

  def song
    ORDINALS
      .zip(GIFTS)
      .each_with_object(EMPTY_SONG.dup, &method(:add_verse_to_song))
      .first
      .join(NEWLINE)
  end

  def add_verse_to_song((day, gift), (song, extra_gifts))
    current_gift = CARDINALS[day] + gift
    todays_gifts = current_gift + extra_gifts.join
    add_current_gift_to_extra_gifts(current_gift, extra_gifts)
    song << verse(day, todays_gifts)
  end
  private_class_method :add_verse_to_song

  def add_current_gift_to_extra_gifts(gift_of_the_day, extra_gifts)
    if extra_gifts.empty?
      extra_gifts << ", and #{gift_of_the_day}"
    else
      extra_gifts.prepend(", #{gift_of_the_day}")
    end
  end
  private_class_method :add_current_gift_to_extra_gifts

  def verse(day, all_gifts)
    format(DECLARATION_OF_RECEIPT, day: day) + all_gifts + PERIOD + NEWLINE
  end
  private_class_method :verse
end

File: elixir/twelve-days/twelve_days.exs
defmodule TwelveDays do
  @cardinals %{
    "first" => "a"
  }
  @gifts %{
    "first" => " Partridge in a Pear Tree"
  }
  @ordinals %{
    1 => "first"
  }
  @period "."

  @doc """
  Given a `number`, return the song's verse for that specific day, including
  all gifts for previous days in the same line.
  """
  @spec verse(number :: integer) :: String.t()
  def verse(number) do
    ordinal =
      @ordinals
      |> Map.get(number)

    ordinal
    |> declaration_of_receipt()
    |> Kernel.<>(Map.get(@cardinals, ordinal))
    |> Kernel.<>(Map.get(@gifts, ordinal))
    |> Kernel.<>(@period)
  end

  @doc """
  Given a `starting_verse` and an `ending_verse`, return the verses for each
  included day, one per line.
  """
  @spec verses(starting_verse :: integer, ending_verse :: integer) :: String.t()
  def verses(starting_verse, ending_verse) do
  end

  @doc """
  Sing all 12 verses, in order, one verse per line.
  """
  @spec sing() :: String.t()
  def sing do
  end

  defp declaration_of_receipt(day) do
    "On the " <> day <> " day of Christmas my true love gave to me: "
  end
end
